Support List.of et al as creating immutable collections

fma1 opened the pull request.

fma1 commented:
@Grundlefleck 
Honestly looking at the code I added, I'm not entirely happy with it. I feel like `checkInvokesImmutableInterfaceMethod()` and `wrapsInUnmodifiable()` still share a bit of logic but I am not sure I can do much about it. In Scala I would just use tuples and pattern matching which would simplify things a lot.

The other thing is that I changed the source and target for `maven-compiler-plugin` in the pom.xml file. Well, IntelliJ complained when I tried to use `List.of()` and I opted to change the language level and IntelliJ changed the pom.xml file.

Grundlefleck commented:
Except for the JDK version change, I would be happy to merge then raise a separate ticket for the suggested refactoring. So if you want the PR merged but don't want to spend much further time, you could drop the version change + test case that drove it, and subsequent PRs can be raised for those. 

OTOH if you want to continue, I'm also happy for review to continue here, prior to merging.

fma1 commented:
@Grundlefleck 
I would like to keep the test there as it's just good practice but I am not sure if that's possible with keeping the JDK version at `1.8`. So I changed `CopyListIntoNewListUsingListOf`'s constructor to `this.immutable = (List<ImmutableExample>) List.class.getMethod("of", Object.class).invoke(null, potentiallyMutatable.get(0));`. That makes the test fail, because it seems that reflection is invoking the method differently:
https://i.imgur.com/sqZWEo3.png

If you're fine with excluding the test case, I suppose I can remove it.

Grundlefleck commented:
I would like the test coverage -- it's definitely good practice as you say. It's just a bit tricky when the source requires different JDK versions. However, it gets less tricky in the ClientOfMutabilityDetector project. What do you think of porting it there, where we have more freedom to switch up JDK versions without impacting users?

Grundlefleck commented:
> That makes the test fail, because it seems that reflection is invoking the method differently:

I'm not surprised, the bytecode of using reflection will be very different, and our naive analysis is not going to be anywhere close to detecting that the reflective call is equivalent to a non-reflective call.

fma1 commented:
@Grundlefleck
Since I don't see any way to include the test in MutabilityDetector, I have removed the test in a new commit. Obviously I think one should be able to use the library in Java 1.8 because for example my current company is still using Java 8 in all their AppServers.

If it's as you say that ClientOfMutabilityDetector has more freedom to up the JDK version, then I think it shouldn't be a problem porting the test to there. I'd obviously have to look at codebase. To be honest, it took me a while to find `allowsCopyingAndWrappingInUmodifiableCollectionTypeIdiom()` in `MutableTypeToFieldCheckerTest`, because I knew I needed a test to confirm that the code I added actually worked in regards to the desired functionality, but I wasn't sure where the test was.

Grundlefleck commented:
I think my inability to use GitHub reviews correctly lead to some confusion. Sorry about that!

fma1 commented:
I see them now. To be honest, I'm more used to using BitBucket than GitHub but it's fine.

fma1 commented on pom.xml:
```diff
@@ -241,8 +241,8 @@
-          <source>1.8</source>
-          <target>1.8</target>
+          <source>9</source>
+          <target>9</target>
```
I think you are commenting on a previous commit as I removed the test case and revert the version change in the pom.xml in commit c333646c7e2424c3f11b2d92749169d4356a61c5 . I also favor option 3 as well.

https://github.com/MutabilityDetector/MutabilityDetector/pull/175/commits/c333646c7e2424c3f11b2d92749169d4356a61c5

fma1 commented on src/main/java/org/mutabilitydetector/checkers/CollectionTypeWrappedInUnmodifiableIdiomOrMadeByImmutableFactoryChecker.java:
```diff
@@ -1,283 +1,312 @@
+import static org.mutabilitydetector.checkers.hint.WrappingHint.NO_HINT;
+import static org.mutabilitydetector.locations.Dotted.dotted;
+
+public class CollectionTypeWrappedInUnmodifiableIdiomOrMadeByImmutableFactoryChecker {
```
Sure that sounds fine to me. I'll make the change.

fma1 commented on src/main/java/org/mutabilitydetector/checkers/CollectionTypeWrappedInUnmodifiableIdiomOrMadeByImmutableFactoryChecker.java:
```diff
@@ -1,283 +1,312 @@
+        return CLASS_NAME_CONVERTER.dotted(typeAssignedToField.getInternalName());
+    }
+
+    private boolean cannotGetMethodNameAndOnClassFromPreviousInvocation(String[] retVal) {
```
Actually, `Optional` seems like a good idea. I'll try that.

fma1 commented on src/main/java/org/mutabilitydetector/checkers/ImmutableCollectionChecker.java:
```diff
@@ -177,41 +179,28 @@ private String typeAssignedToField() {
+        return getLastMethodInsnNode().map(previousInvocation ->
+                        Configuration.INSTANCE.UNMODIFIABLE_METHOD_OWNER.equals(CLASS_NAME_CONVERTER.dotted(previousInvocation.owner)) &&
+                        Configuration.INSTANCE.FIELD_TYPE_TO_UNMODIFIABLE_METHOD.get(typeAssignedToField()).equals(previousInvocation.name))
+                    .orElse(false);
```
So I know you wanted me to use a method reference. I've mainly used them when IntelliJ has suggested them for replacing a lambda. I can't use it here because of both `onClass` and `methodName` being required for the boolean expression.

Hear me out here. What I might be able to do is declare something like this:
`private static final Function<MethodInsnNode, Boolean> someFuncName = previousInvocation -> ...` And then I could just supply that function to `Optional#map()` in both `checkInvokesImmutableFactoryMethod()` and `wrapsInUmodifiable()`.

I am thinking that what we could do is merge `FIELD_TYPE_TO_UNMODIFIABLE_METHOD` and `FIELD_TYPE_TO_IMMUTABLE_METHOD`. I'm also thinking maybe I can add a dummy key-value pair to the merged map with `java.util.Collections` as the key in that dummy key-value pair. Because for the Java 9 Immutable factory methods, the `owner` depends on whether it's a `List`, a `Set` or a `Map`. Hmm, do you see any problems with using `containsKey()` and that dummy KV pair for the unmodifiable collections?

I'm only suggesting this because you suggested a method reference. In order to supply one lambda for both methods, there needs to be generalization between the two methods.

![unmodifiableList](https://user-images.githubusercontent.com/15847948/86194794-71728e00-bb1d-11ea-93cd-bd3900060bbc.png)

fma1 replied:
Actually thinking about this, if I could generalize then `wrapInUnmodifiable()` and `checkInvokesImmutableFactoryMethod()` could be consolidated into one method. Also thinking about this, a lambda function wouldn't be the best option because the `Boolean` result is based on the state of the instance.

fma1 replied:
But maybe it's better to leave them as 2 methods. Because there's more checking after `wrapsInUnmodifiable()` is finally called.

Grundlefleck replied:
@fma1 what you have now looks good to me. I had in mind 2x public methods + 2x private methods used as a method reference (of type `Function<MethodInsnNode, Boolean>` as you say). But I think 2x public methods, each with an inline lambda, is better.

fma1 commented:
I made the modifications requested but looking for further clarification about the method reference (well lambda function).

fma1 commented:
Just to make sure, the latest commit is commit d45460ec25c6bee145ddb40ba29db26bb80af2dd .

Grundlefleck commented:
I think this is good to go. Good work, and thanks for your persistence @fma1, particularly with my delayed feedback. 

Are you interested in issuing the PR to provide test coverage via ClientOfMutabilityDetector? I'd welcome a further PR on that repo for sure!

fma1 commented:
@Grundlefleck 
Sure, I'd be happy to take a look at ClientOfMutabilityDetector to add that test.
